django/contrib/auth/password_validation.py

```diff
@@ -0,0 +1,134 @@
+    """
+    Returns a list of all help texts of all configured validators.
+    """
+    help_texts = []
+    for validator in get_password_validators():
+        help_texts.append(validator.get_help_text())
+    return help_texts
+
+
+def password_validators_help_text_html():
+    """
+    Returns an HTML string with all help texts of all configured validators in an <ul>.
+    """
+    help_texts = password_validators_help_texts()
+    return '<ul>%s</ul>' % unordered_list(help_texts)
+
+
+class MinimumLengthValidator(object):
+    """
+    Validates whether the password has a minimum length.
+    """
+    def __init__(self, min_length=8):
+        self.min_length = min_length
+
+    def validate(self, password, user=None):
+        if len(password) < self.min_length:
+            return _("This password is too short.")
+
+    def get_help_text(self):
+        return _("Your password must contain at least %(min_length)d characters.") % {'min_length': self.min_length}
+
+
+class UserAttributeSimilarityValidator(object):
+    """
+    Validates whether the password is sufficiently different from the user's attributes.
+
+    If no specific attributes are provided, this looks at a sensible list of defaults.
+    Attributes that don't exist are gracefully ignored. Comparison is made to not only
+    the full attribute value, but also it's components, so that e.g. a password is validated
+    against either part of an e-mail address, as well as the full address.
+    """
+    def __init__(self, user_attributes=None, max_similarity=0.7):
+        self.user_attributes = user_attributes if user_attributes else ('username', 'first_name', 'last_name', 'email')
+        self.max_similarity = max_similarity
+
+    def validate(self, password, user=None):
+        if not user:
+            return
+
+        for attribute_name in self.user_attributes:
+            value = getattr(user, attribute_name, None)
+            if not value or not isinstance(value, string_types):
+                continue
+            value_parts = re.split('\W+', value) + [value]
+            for value_part in value_parts:
+                if SequenceMatcher(a=password.lower(), b=value_part.lower()).quick_ratio() > self.max_similarity:
+                    verbose_name = force_text(user._meta.get_field(attribute_name).verbose_name)
+                    return _("The password is too similar to the %s." % verbose_name)
+
+    def get_help_text(self):
+        return _("Your password must not be too similar to your other personal information.")
+
+
+class CommonPasswordValidator(object):
+    """
+    Validates whether the password is a common password.
+
+    The password is rejected if it occurs in a list of 1000 common passwords.
+    The list used was created by Mark Burnett: https://xato.net/passwords/more-top-worst-passwords/
+    """
+    def __init__(self):
+        common_passwords_file = os.path.dirname(os.path.realpath(__file__)) + '/common-passwords.txt'
```

Comment:
With the discussion leaning towards shipping a gzipped list, so that the Django codebase doesn't contain so many bad words in cleartext, I guess the python option is no longer advantageous. Making it customisable would be useful though. I'm not in favour of shipping a separate list, due to the increased effort required from our users.